### DISCUSSION TITLE
Retrieving group members?

### DISCUSSION THREAD

**Discussion by LegendaryB** [pnp/pnpcore#764] [Q&A] [answered]
Hey,

our application retrieves a list of IFieldUserValues. Some of them are groups and some of them normal users. The requirement is to resolve the groups to a list of users (mail addresses). How can we do that with PnP? Or is it only possible via a custom MS Graph API call? The groups are not contained in SiteGroups or the Associated groups.

```csharp
var item = await list.Items.GetByIdAsync(id);
var fieldUserValueCollection = item.Values["Gruppen"] as IFieldValueCollection;

foreach (IFieldUserValue userValue in fieldUserValueCollection.Values)
{
    var user = await web.GetUserByIdAsync(userValue.LookupId);

    if (IsGroup(user)) {
       // Resolve users
    }
}

private bool IsGroup(ISharePointUser user)
{
    return user.PrincipalType == PrincipalType.DistributionList ||
               user.PrincipalType == PrincipalType.SecurityGroup ||
               user.PrincipalType == PrincipalType.SharePointGroup;
}
```

**Comment by jansenbe** [comment, marked as answer]
@LegendaryB : for SharePoint group you'll be able to load the respective members and for those who are real users you can get the email/upn. For the others (Azure AD security groups / distribution lists) you would need to indeed use custom Graph APIs. Do note that in SharePoint one could also directly have permissioned an Azure AD group without adding it to a SharePoint group first.

A common problem with this is the group nesting, especially with Azure AD groups...image someone granting the "All Employees" group access, that might lead to a tree of nested groups.

https://docs.microsoft.com/en-us/graph/api/group-list-transitivemembers?view=graph-rest-1.0&tabs=http seems to be good start (did not try this) and see https://pnp.github.io/pnpcore/using-the-sdk/basics-customapirequests.html#making-a-custom-microsoft-graph-request on how to make custom graph requests with PnP Core SDK. At the bottom of that page there's also details on how to add custom request headers (at that might be needed for that API call).

https://github.com/baswijdenes/BasWijdenes/blob/master/Get-AzureADNestedGroupMember.ps1 is a PS example showing how to recursively handle nested groups.

**Reply by LegendaryB** [reply]
@jansenbe Thanks for your detailed answer. I guessed so. I will take a look at the appropriate MS Graph API. I already know how to do the custom request stuff with pnpcore. After some work with pnpcore I got to know the framework :D

Will mark this as answer :)

**Reply by LegendaryB** [reply]
@jansenbe 1 question left, other stuff is clear. A group model class is nowhere available? So that I can JsonConvert the response directly in it? I guess no, because the API must be custom called? Just wanted to make sure that I don't miss that before writing the model myself :D

**Edit:** nvm, forgot about the MS graph client^^
